pkg/frontend/querymiddleware/step_align.go

```diff
@@ -48,33 +54,35 @@ func (s *stepAlignMiddleware) Do(ctx context.Context, r MetricsQueryRequest) (Re
 		return s.next.Do(ctx, r)
 	}
 
-	if validation.AllTrueBooleansPerTenant(tenants, s.limits.AlignQueriesWithStep) {
-		start := (r.GetStart() / r.GetStep()) * r.GetStep()
-		end := (r.GetEnd() / r.GetStep()) * r.GetStep()
-
-		if start != r.GetStart() || end != r.GetEnd() {
-			for _, id := range tenants {
-				s.aligned.WithLabelValues(id).Inc()
-			}
-
-			spanlog := spanlogger.FromContext(ctx, s.logger)
-			spanlog.DebugLog(
-				"msg", "query start or end has been adjusted to be step-aligned",
-				spanlogger.TenantIDsTagName, tenants,
-				"original_start", r.GetStart(),
-				"original_end", r.GetEnd(),
-				"adjusted_start", start,
-				"adjusted_end", end,
-				"step", r.GetStep(),
-			)
-
-			updatedReq, err := r.WithStartEnd(start, end)
-			if err != nil {
-				return nil, err
-			}
-
-			return s.next.Do(ctx, updatedReq)
+	start := (r.GetStart() / r.GetStep()) * r.GetStep()
+	end := (r.GetEnd() / r.GetStep()) * r.GetStep()
+	stepAligned := start == r.GetStart() && end == r.GetEnd()
```

Comment:
[nit] Might be clearer to do something like `if stepAligned { return s.next.Do(ctx, r) }` here to avoid the need for the conditions below

Reply:
I've reversed the conditionals so we return early in all cases unless we are adjusting the start/end.